File: internal/handler/refine.go
package handler

import (
	"errors"
	"net/http"
	"strings"
	"time"

	"changkun.de/wallfacer/internal/logger"
	"changkun.de/wallfacer/internal/store"
	"github.com/google/uuid"
)

// StartRefinementRequest is the optional body for POST /api/tasks/{id}/refine.
type StartRefinementRequest struct {
	UserInstructions string `json:"user_instructions"`
}

// StartRefinement starts a sandbox-based refinement run for a backlog task.
// The sandbox agent explores the codebase and produces a detailed implementation spec.
// POST /api/tasks/{id}/refine
func (h *Handler) StartRefinement(w http.ResponseWriter, r *http.Request, id uuid.UUID) {
	task, err := h.store.GetTask(r.Context(), id)
	if err != nil {
		http.Error(w, "task not found", http.StatusNotFound)
		return
	}
	if task.Status != "backlog" {
		http.Error(w, "task is not in backlog", http.StatusBadRequest)
		return
	}

	var req StartRefinementRequest
	if r.ContentLength > 0 {
		// Body is optional; ignore decode errors (empty or malformed body → no instructions).
		decodeOptionalJSONBody(r, &req)
	}

	job := &store.RefinementJob{
		ID:        uuid.New().String(),
		CreatedAt: time.Now(),
		Status:    "running",
		Source:    "runner",
	}
	if err := h.store.StartRefinementJobIfIdle(r.Context(), id, job); err != nil {
		if errors.Is(err, store.ErrRefinementAlreadyRunning) {
			http.Error(w, "refinement already running", http.StatusConflict)
			return
		}
		logger.Handler.Error("start refinement: start job", "task", id, "error", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	h.runner.RunRefinementBackground(id, strings.TrimSpace(req.UserInstructions))

	updated, err := h.store.GetTask(r.Context(), id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	writeJSON(w, http.StatusAccepted, updated)
}

// CancelRefinement stops a running sandbox refinement by killing the container.
// DELETE /api/tasks/{id}/refine
func (h *Handler) CancelRefinement(w http.ResponseWriter, r *http.Request, id uuid.UUID) {
	task, err := h.store.GetTask(r.Context(), id)
	if err != nil {
		http.Error(w, "task not found", http.StatusNotFound)
		return
	}
	if task.CurrentRefinement == nil || task.CurrentRefinement.Status != "running" {
		http.Error(w, "no refinement running", http.StatusBadRequest)
		return
	}

	h.runner.KillRefineContainer(id)

	// Mark as failed (cancelled).
	task.CurrentRefinement.Status = "failed"
	task.CurrentRefinement.Error = "cancelled by user"
	if err := h.store.UpdateRefinementJob(r.Context(), id, task.CurrentRefinement); err != nil {
		logger.Handler.Error("cancel refinement: update job", "task", id, "error", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	updated, err := h.store.GetTask(r.Context(), id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	writeJSON(w, http.StatusOK, updated)
}

// RefineDissmiss clears a completed refinement result without applying it.
// The task prompt remains unchanged; the CurrentRefinement field is cleared so
// the task can be started normally again.
// POST /api/tasks/{id}/refine/dismiss
func (h *Handler) RefineDismiss(w http.ResponseWriter, r *http.Request, id uuid.UUID) {
	task, err := h.store.GetTask(r.Context(), id)
	if err != nil {
		http.Error(w, "task not found", http.StatusNotFound)
		return
	}
	if task.Status != "backlog" {
		http.Error(w, "task is not in backlog", http.StatusBadRequest)
		return
	}
	if task.CurrentRefinement == nil || task.CurrentRefinement.Status != "done" {
		http.Error(w, "no completed refinement to dismiss", http.StatusBadRequest)
		return
	}

	if err := h.store.DismissRefinement(r.Context(), id); err != nil {
		logger.Handler.Error("dismiss refinement", "task", id, "error", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	updated, err := h.store.GetTask(r.Context(), id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	writeJSON(w, http.StatusOK, updated)
}

// RefineApplyRequest is the body for POST /api/tasks/{id}/refine/apply.
type RefineApplyRequest struct {
	Prompt string `json:"prompt"`
}

// RefineApply persists the refined prompt, recording the session in
// RefineSessions and moving the old prompt to PromptHistory.
// POST /api/tasks/{id}/refine/apply
func (h *Handler) RefineApply(w http.ResponseWriter, r *http.Request, id uuid.UUID) {
	task, err := h.store.GetTask(r.Context(), id)
	if err != nil {
		http.Error(w, "task not found", http.StatusNotFound)
		return
	}
	if task.Status != "backlog" {
		http.Error(w, "task is not in backlog", http.StatusBadRequest)
		return
	}

	var req RefineApplyRequest
	if !decodeJSONBody(w, r, &req) {
		return
	}
	if strings.TrimSpace(req.Prompt) == "" {
		http.Error(w, "prompt is required", http.StatusBadRequest)
		return
	}

	// Build a session recording what the sandbox produced vs what the user applied.
	sandboxResult := ""
	if task.CurrentRefinement != nil {
		sandboxResult = task.CurrentRefinement.Result
	}
	session := store.RefinementSession{
		ID:          uuid.New().String(),
		CreatedAt:   time.Now(),
		StartPrompt: task.Prompt,
		Result:      sandboxResult,
	}

	if err := h.store.ApplyRefinement(r.Context(), id, req.Prompt, session); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	// Regenerate title for the updated prompt.
	h.runner.GenerateTitleBackground(id, req.Prompt)

	updated, err := h.store.GetTask(r.Context(), id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	writeJSON(w, http.StatusOK, updated)
}

File: ui/js/refine.js
// --- Task refinement via sandbox agent ---
//
// Flow:
//   1. User clicks "Start Refinement" → POST /api/tasks/{id}/refine
//   2. Sandbox agent runs read-only, produces a spec
//   3. Live logs stream via GET /api/tasks/{id}/refine/logs
//   4. On completion, result appears in an editable textarea
//   5. User optionally edits and clicks "Apply as Prompt"

let refineTaskId = null;
let refineLogsAbort = null; // AbortController for the live log stream

// updateRefineUI re-renders the refinement panel based on the current task state.
// Called whenever the modal opens or the task object changes via SSE.
function updateRefineUI(task) {
  if (!task || task.status !== 'backlog') return;

  const job = task.current_refinement;

  const startBtn   = document.getElementById('refine-start-btn');
  const cancelBtn  = document.getElementById('refine-cancel-btn');
  const running    = document.getElementById('refine-running');
  const resultSec  = document.getElementById('refine-result-section');
  const errorSec   = document.getElementById('refine-error-section');
  const resultTA   = document.getElementById('refine-result-prompt');
  const errorMsg   = document.getElementById('refine-error-msg');

  // Determine UI state from job status.
  if (!job) {
    showRefineIdle(startBtn, cancelBtn, running, resultSec, errorSec);
    return;
  }

  if (job.status === 'running') {
    startBtn.classList.add('hidden');
    cancelBtn.classList.remove('hidden');
    running.classList.remove('hidden');
    resultSec.classList.add('hidden');
    errorSec.classList.add('hidden');
    const idleDesc = document.getElementById('refine-idle-desc');
    if (idleDesc) idleDesc.classList.add('hidden');
    const instrSec = document.getElementById('refine-instructions-section');
    if (instrSec) instrSec.classList.add('hidden');

    // Attach log stream if this is the active task and not already streaming.
    if (refineTaskId === task.id && !refineLogsAbort) {
      startRefineLogStream(task.id);
    }
    return;
  }

  if (job.status === 'done') {
    startBtn.classList.remove('hidden');
    cancelBtn.classList.add('hidden');
    running.classList.add('hidden');
    resultSec.classList.remove('hidden');
    errorSec.classList.add('hidden');
    const idleDesc = document.getElementById('refine-idle-desc');
    if (idleDesc) idleDesc.classList.add('hidden');
    const instrSec = document.getElementById('refine-instructions-section');
    if (instrSec) instrSec.classList.add('hidden');
    stopRefineLogStream();

    // Only populate the textarea if it is empty or this is the first population.
    if (resultTA.dataset.jobId !== job.id) {
      resultTA.value = job.result || '';
      resultTA.dataset.jobId = job.id;
    }

    // Show the dismiss button so user can skip applying the refinement.
    const dismissBtn = document.getElementById('refine-dismiss-btn');
    if (dismissBtn) dismissBtn.classList.remove('hidden');
    return;
  }

  if (job.status === 'failed') {
    showRefineIdle(startBtn, cancelBtn, running, resultSec, errorSec);
    errorSec.classList.remove('hidden');
    errorMsg.textContent = 'Refinement failed: ' + (job.error || 'unknown error');
    stopRefineLogStream();
    return;
  }
}

function showRefineIdle(startBtn, cancelBtn, running, resultSec, errorSec) {
  startBtn.classList.remove('hidden');
  cancelBtn.classList.add('hidden');
  running.classList.add('hidden');
  resultSec.classList.add('hidden');
  errorSec.classList.add('hidden');
  const idleDesc = document.getElementById('refine-idle-desc');
  if (idleDesc) idleDesc.classList.remove('hidden');
  const instrSec = document.getElementById('refine-instructions-section');
  if (instrSec) instrSec.classList.remove('hidden');
  const dismissBtn = document.getElementById('refine-dismiss-btn');
  if (dismissBtn) dismissBtn.classList.add('hidden');
}

// startRefinement is called by the "Start" button.
async function startRefinement() {
  if (!currentTaskId) return;

  // If refinement is already running, ignore the click.
  const currentTask = tasks.find(t => t.id === currentTaskId);
  if (currentTask && currentTask.current_refinement && currentTask.current_refinement.status === 'running') return;

  refineTaskId = currentTaskId;

  // Clear prior log output and reset mode.
  refineRawLogBuffer = '';
  refineLogsMode = 'pretty';
  setRefineLogsMode('pretty');
  const logsEl = document.getElementById('refine-logs');
  if (logsEl) logsEl.innerHTML = '';

  // Clear prior result textarea job-id so result gets populated fresh.
  const resultTA = document.getElementById('refine-result-prompt');
  if (resultTA) delete resultTA.dataset.jobId;

  try {
    const userInstructions = document.getElementById('refine-user-instructions')?.value.trim() || '';
    await api(task(currentTaskId).refine(), {
      method: 'POST',
      body: JSON.stringify({ user_instructions: userInstructions }),
    });
    // SSE task stream will push the updated task; updateRefineUI handles the rest.
  } catch (e) {
    const errorSec = document.getElementById('refine-error-section');
    const errorMsg = document.getElementById('refine-error-msg');
    if (errorSec) errorSec.classList.remove('hidden');
    if (errorMsg) errorMsg.textContent = 'Failed to start refinement: ' + e.message;
  }
}

// cancelRefinement is called by the "Cancel" button.
async function cancelRefinement() {
  if (!refineTaskId) return;
  stopRefineLogStream();
  try {
    await api(task(refineTaskId).refine(), { method: 'DELETE' });
  } catch (e) {
    // Ignore — SSE will reflect the updated state.
  }
}

// --- Refine log render scheduler ---
// Coalesces rapid chunk arrivals into a single paint per animation frame,
// mirroring the scheduleLogRender() pattern used by modal-logs.js.
let _refineLogRenderPending = false;
function scheduleRefineLogRender() {
  if (_refineLogRenderPending) return;
  _refineLogRenderPending = true;
  requestAnimationFrame(function() {
    _refineLogRenderPending = false;
    renderRefineLogs();
  });
}

// renderRefineLogs re-renders the refine log area from refineRawLogBuffer.
function renderRefineLogs() {
  const logsEl = document.getElementById('refine-logs');
  if (!logsEl) return;
  // Read scroll state before mutating the DOM to avoid a forced synchronous layout
  // between the read and the subsequent innerHTML write.
  const atBottom = logsEl.scrollHeight - logsEl.scrollTop - logsEl.clientHeight < 80;
  if (refineLogsMode === 'pretty') {
    logsEl.innerHTML = renderPrettyLogs(refineRawLogBuffer);
  } else {
    logsEl.textContent = refineRawLogBuffer.replace(/\x1b\[[0-9;]*[a-zA-Z]/g, '');
  }
  if (atBottom) {
    // Defer the scroll-to-bottom to the next frame so the browser can batch
    // the layout triggered by the innerHTML write with the scroll update.
    requestAnimationFrame(function() {
      logsEl.scrollTop = logsEl.scrollHeight;
    });
  }
}

// setRefineLogsMode switches between 'pretty' and 'raw' and re-renders.
function setRefineLogsMode(mode) {
  refineLogsMode = mode;
  ['pretty', 'raw'].forEach(function(m) {
    const tab = document.getElementById('refine-logs-tab-' + m);
    if (tab) tab.classList.toggle('active', m === mode);
  });
  renderRefineLogs();
}

// startRefineLogStream opens a streaming fetch to the refine/logs endpoint
// and accumulates chunks into refineRawLogBuffer for pretty/raw rendering.
function startRefineLogStream(taskId) {
  if (refineLogsAbort) return; // already streaming
  refineLogsAbort = new AbortController();

  const decoder = new TextDecoder();

  fetch(task(taskId).refineLogs(), { signal: refineLogsAbort.signal })
    .then(async resp => {
      if (resp.status === 204) {
        // Container already done.
        refineLogsAbort = null;
        return;
      }
      if (!resp.ok || !resp.body) {
        refineLogsAbort = null;
        return;
      }
      const reader = resp.body.getReader();
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        refineRawLogBuffer += decoder.decode(value, { stream: true });
        scheduleRefineLogRender();
      }
      refineLogsAbort = null;
    })
    .catch(err => {
      if (err.name !== 'AbortError') {
        console.warn('refine log stream error:', err);
      }
      refineLogsAbort = null;
    });
}

// stopRefineLogStream aborts any active log stream.
function stopRefineLogStream() {
  if (refineLogsAbort) {
    refineLogsAbort.abort();
    refineLogsAbort = null;
  }
}

// resetRefinePanel resets the panel state when the modal closes or switches tasks.
function resetRefinePanel() {
  refineTaskId = null;
  stopRefineLogStream();
  // Reset all sub-elements individually to avoid errors when elements are absent.
  const startBtn  = document.getElementById('refine-start-btn');
  const cancelBtn = document.getElementById('refine-cancel-btn');
  const running   = document.getElementById('refine-running');
  const resultSec = document.getElementById('refine-result-section');
  const errorSec  = document.getElementById('refine-error-section');
  if (startBtn)  startBtn.classList.remove('hidden');
  if (cancelBtn) cancelBtn.classList.add('hidden');
  if (running)   running.classList.add('hidden');
  if (resultSec) resultSec.classList.add('hidden');
  if (errorSec)  errorSec.classList.add('hidden');
  const idleDesc = document.getElementById('refine-idle-desc');
  if (idleDesc) idleDesc.classList.remove('hidden');
  const instrSec = document.getElementById('refine-instructions-section');
  if (instrSec) instrSec.classList.remove('hidden');
  const instrTA = document.getElementById('refine-user-instructions');
  if (instrTA) instrTA.value = '';
  const resultTA = document.getElementById('refine-result-prompt');
  if (resultTA) delete resultTA.dataset.jobId;
  const dismissBtn = document.getElementById('refine-dismiss-btn');
  if (dismissBtn) dismissBtn.classList.add('hidden');
  refineRawLogBuffer = '';
  refineLogsMode = 'pretty';
  ['pretty', 'raw'].forEach(function(m) {
    const tab = document.getElementById('refine-logs-tab-' + m);
    if (tab) tab.classList.toggle('active', m === 'pretty');
  });
  const logsEl = document.getElementById('refine-logs');
  if (logsEl) logsEl.innerHTML = '';
}

// dismissRefinement clears the refinement result without applying it, allowing
// the task to be started with the original prompt.
async function dismissRefinement() {
  if (!currentTaskId) return;
  try {
    await api(task(currentTaskId).refineDismiss(), { method: 'POST' });
    closeModal();
    fetchTasks();
  } catch (e) {
    showAlert('Error dismissing refinement: ' + e.message);
  }
}

// applyRefinement POSTs the (possibly edited) spec as the new task prompt.
async function applyRefinement() {
  if (!currentTaskId) return;
  const resultTA = document.getElementById('refine-result-prompt');
  const newPrompt = resultTA ? resultTA.value.trim() : '';
  if (!newPrompt) {
    showAlert('The refined prompt cannot be empty.');
    return;
  }

  try {
    // Save settings changes (sandbox, timeout, mount_worktrees) alongside the apply.
    const sandbox = document.getElementById('modal-edit-sandbox')?.value || '';
    const sandboxByActivity = collectSandboxByActivity('modal-edit-sandbox-');
    const timeout = parseInt(document.getElementById('modal-edit-timeout')?.value, 10) || DEFAULT_TASK_TIMEOUT;
    const mountWorktrees = document.getElementById('modal-edit-mount-worktrees')?.checked || false;
    await api(task(currentTaskId).update(), {
      method: 'PATCH',
      body: JSON.stringify({ sandbox, sandbox_by_activity: sandboxByActivity, timeout, mount_worktrees: mountWorktrees }),
    });

    await api(task(currentTaskId).refineApply(), {
      method: 'POST',
      body: JSON.stringify({ prompt: newPrompt }),
    });

    await fetchTasks();
    openModal(currentTaskId);
  } catch (e) {
    showAlert('Error applying refinement: ' + e.message);
  }
}

// renderRefineHistory populates the history section from task.refine_sessions.
function renderRefineHistory(task) {
  const section = document.getElementById('refine-history-section');
  const list = document.getElementById('refine-history-list');
  const sessions = (task.refine_sessions || []);
  if (sessions.length === 0) {
    section.classList.add('hidden');
    return;
  }
  section.classList.remove('hidden');
  list.innerHTML = sessions.map((s, i) => {
    const date = new Date(s.created_at).toLocaleString();
    const previewPrompt = s.start_prompt || '';
    const resultPrompt = s.result_prompt || '';
    const sandboxResult = s.result || '';
    return `<details class="refine-history-entry">
      <summary class="refine-history-summary">
        <span class="text-xs text-v-muted">#${i + 1} · ${escapeHtml(date)}</span>
      </summary>
      <div style="padding:8px 0 0 0;">
        <div class="text-xs text-v-muted" style="margin-bottom:4px;">Starting prompt:</div>
        <pre class="code-block text-xs" style="white-space:pre-wrap;word-break:break-word;opacity:0.7;">${escapeHtml(previewPrompt)}</pre>
        ${sandboxResult ? `
        <details style="margin-top:8px;">
          <summary class="text-xs text-v-muted" style="cursor:pointer;">Sandbox spec (before editing)</summary>
          <pre class="code-block text-xs" style="white-space:pre-wrap;word-break:break-word;margin-top:4px;opacity:0.8;">${escapeHtml(sandboxResult)}</pre>
        </details>
        ` : ''}
        ${resultPrompt ? `
        <div class="text-xs text-v-muted" style="margin-top:8px;margin-bottom:4px;">Applied prompt:</div>
        <pre class="code-block text-xs" style="white-space:pre-wrap;word-break:break-word;">${escapeHtml(resultPrompt)}</pre>
        <button class="btn btn-ghost text-xs" style="margin-top:6px;" onclick="revertToHistoryPrompt(${i})">Revert to this version</button>
        ` : ''}
      </div>
    </details>`;
  }).join('');
}

// revertToHistoryPrompt loads a previous session's applied prompt into the result textarea.
function revertToHistoryPrompt(sessionIndex) {
  const currentTask = tasks.find(t => t.id === currentTaskId);
  if (!currentTask || !currentTask.refine_sessions) return;
  const session = currentTask.refine_sessions[sessionIndex];
  if (!session || !session.result_prompt) return;

  const resultTA = document.getElementById('refine-result-prompt');
  if (!resultTA) return;

  // Show the result section so the user can apply.
  document.getElementById('refine-result-section').classList.remove('hidden');
  resultTA.value = session.result_prompt;
  delete resultTA.dataset.jobId;
}
